--- a/arch/arm64/boot/Makefile	2025-03-05 21:35:34.252407159 +0800
+++ b/arch/arm64/boot/Makefile	2025-03-05 21:36:13.846733684 +0800
@@ -28,6 +28,8 @@
 endif
 DTB_OBJS := $(shell find $(obj)/dts/ -name \*.dtb)
 
+DTBO_OBJS := $(shell find $(obj)/dts/ -name \*.dtbo)
+
 # Add RTIC DTB to the DTB list if RTIC MPGen is enabled
 # Note, we keep this for compatibility with
 # BUILD_ARM64_APPENDED_DTB_IMAGE targets.

--- a/arch/arm64/Makefile	2025-03-05 21:39:20.314858191 +0800
+++ b/arch/arm64/Makefile	2025-03-05 21:38:41.435267334 +0800
@@ -180,6 +180,13 @@
 Image.gz-dtb: vmlinux scripts dtbs Image.gz
 	$(Q)$(MAKE) $(build)=$(boot) $(boot)/$@
 
+ifeq ($(CONFIG_BUILD_ARM64_DT_OVERLAY),y)
+$(KBUILD_DTBO_IMG): dtbs
+	$(Q)$(MAKE) $(build)=$(boot) $(boot)/$@
+
+all: $(KBUILD_DTBO_IMG)
+endif
+
 PHONY += vdso_install
 vdso_install:
 	$(Q)$(MAKE) $(build)=arch/arm64/kernel/vdso $@

